smarts/core/lanepoints.py

```diff
@@ -484,15 +608,25 @@ def paths_starting_at_lanepoint(
             for path in lanepoint_paths:
                 branching_paths = []
                 for next_lp in path[-1].nexts:
-                    # TODO: This could be a problem. What about internal lanes?
+                    # TODO: This could be a problem for SUMO. What about internal lanes?
                     # Filter only the edges we're interested in
-                    edge_id = next_lp.lp.lane.road.road_id
+                    next_lane = next_lp.lp.lane
+                    edge_id = next_lane.road.road_id
                     if filter_edge_ids and edge_id not in filter_edge_ids:
                         continue
+                    if (
+                        filter_edge_ids
+                        and edge_id != filter_edge_ids[-1]
+                        and all(
+                            out_lane.road.road_id not in filter_edge_ids
+                            for out_lane in next_lane.outgoing_lanes
+                        )
+                    ):
+                        continue
```

Comment:
This was added to prevent path generation along lanes which don't lie on the roads not in the `Route` passed. Since for `OpenDRIVE` we add all the lanes of the outgoing road as the next possible lane the vehicle go to (allow lane changes when road change),

```python
                outgoing_roads_added = []
                for out_lane in curr_lane.outgoing_lanes:
                    if out_lane.is_drivable:
                        lane_queue.put((out_lane, curr_lanepoint))
                    outgoing_road = out_lane.road
                    if out_lane.road not in outgoing_roads_added:
                        outgoing_roads_added.append(outgoing_road)
                for outgoing_road in outgoing_roads_added:
                    for next_lane in outgoing_road.lanes:
                        if (
                            next_lane.is_drivable
                            and next_lane not in curr_lane.outgoing_lanes
                        ):
                            lane_queue.put((next_lane, curr_lanepoint))

```

unlike `SUMO` which only allows the to go into outgoing lanes from a given lane.
```python
                shape_lanepoints.append(last_linked_lanepoint)
                curr_lanepoint.nexts.append(last_linked_lanepoint)
                curr_lanepoint = last_linked_lanepoint

                for out_connection in lane.getOutgoing():
                    out_lane = out_connection.getToLane()

                    # Use internal lanes of junctions (if we're at a junction)
                    via_lane_id = out_connection.getViaLaneID()
                    if via_lane_id:
                        out_lane = road_map._graph.getLane(via_lane_id)

                    lane_queue.put((out_lane, curr_lanepoint))
```
The connections can cause the path generation to generate a path which go into road not in Route.
So the heuristic above prevents that by doing a forward check.